plugins/modem-manager/fu-mm-device.c

```diff
@@ -1279,10 +1217,109 @@ fu_mm_device_write_firmware_mbim_qdu(FuDevice *device,
 #endif /* MM_CHECK_VERSION(1,17,1) && MBIM_CHECK_VERSION(1,25,3) */
 
 #if MM_CHECK_VERSION(1, 17, 2)
+static gboolean
+fu_mm_find_device_file(FuDevice *device, gpointer userdata, GError **error)
+{
+	FuMmDevice *self = FU_MM_DEVICE(device);
+	const gchar *subsystem = (const gchar *)userdata;
+
+	return fu_mm_utils_find_device_file(fu_device_get_physical_id(FU_DEVICE(self)),
+					    subsystem,
+					    &self->port_edl,
+					    error);
+}
+
+static gboolean
+fu_mm_device_find_edl_port(FuDevice *device, const gchar *subsystem, GError **error)
+{
+	FuMmDevice *self = FU_MM_DEVICE(device);
+
+	g_clear_pointer(&self->port_edl, g_free);
+
+	return fu_device_retry_full(device,
+				    fu_mm_find_device_file,
+				    30,
+				    250,
+				    (gpointer)subsystem,
+				    error);
+}
+
+static gboolean
+fu_mm_device_qcdm_switch_to_edl(FuDevice *device, gpointer userdata, GError **error)
+{
+	FuMmDevice *self = FU_MM_DEVICE(device);
+	g_autoptr(FuDeviceLocker) locker = NULL;
+	static const guint8 emergency_download[] = {0x4b, 0x65, 0x01, 0x00, 0x54, 0x0f, 0x7e};
+
+	locker = fu_device_locker_new_full(self,
+					   (FuDeviceLockerFunc)fu_mm_device_io_open_qcdm,
+					   (FuDeviceLockerFunc)fu_mm_device_io_close,
+					   error);
+
+	if (locker == NULL) {
+		if (g_error_matches(*error, FWUPD_ERROR, FWUPD_ERROR_INVALID_FILE)) {
+			g_clear_error(error);
```

Comment:
you need to use error_local here and then do `g_propagate_error(error, g_steal_pointer(&error_local));` if it doesn't match. In general you can't assume that the `GError **error` is non-NULL.

Reply:
thanks for the tip, fixed it in the code now